pkg/cloud-controller-manager/loadbalancer.go

```diff
@@ -257,3 +309,32 @@ func waitForIP(callback func() (runtime.Object, string, error)) (runtime.Object,
 		}
 	}
 }
+
+func (l *LoadBalancerManager) checkPortOverlap(primary, secondary *v1.Service) error {
+	if len(primary.Status.LoadBalancer.Ingress) == 0 {
+		return fmt.Errorf("primary service %s/%s has no ingress IP", primary.Namespace, primary.Name)
+	}
+
+	svcs, err := l.localSvcCache.List(metav1.NamespaceAll, labels.NewSelector())
+	if err != nil {
+		return fmt.Errorf("list service failed: %w", err)
+	}
+	// port map is used to check if the port is already used by the primary service and its secondary services
+	portMap := make(map[int32]string)
+	for _, svc := range svcs {
+		if svc.UID == secondary.UID || len(svc.Status.LoadBalancer.Ingress) == 0 || svc.Status.LoadBalancer.Ingress[0].IP != primary.Status.LoadBalancer.Ingress[0].IP {
+			continue
+		}
+		for _, port := range svc.Spec.Ports {
+			portMap[port.Port] = svc.Namespace + "/" + svc.Name
```

Comment:
we may avoid building a map,
in L329, put the loop of L333~L337

or in reverse order, build a map of secondary.Spec.Ports (it should be smaller and limited), then in loop of svcs check if the port hit in this small map.